### PR TITLE
[collections] Add a RangeSet collection.

### PR DESCRIPTION
**Pull request by garretrieger**
This stores a disjoint list of ranges with automatica merging of overlapping and adjacent ranges.

### CODE REVIEW

File: read-fonts/src/collections/range_set/mod.rs [hunk 1]
```diff
@@ -0,0 +1,246 @@
...
+
+#[derive(Default)]
+pub struct RangeSet<T> {
+    // an entry in the map ranges[a] = b implies there is an range [a, b] (inclusive) in this set.
+    ranges: BTreeMap<T, T>,
+}
+
+pub trait Sequence<T> {
+    fn next(&self) -> Option<T>;
+}
```
[the remark this answers is not in the document]
Comment (cmyr) [reply read-fonts/src/collections/range_set/mod.rs:21]: Okay, so after a bit deeper digging:

- what we really care about here is the idea of a totally ordered set, and what we need is the ability to determine if one value immediately follows another in the ordering. (I'm sure there's a math name for this, but I don't know it)
- Really we get the majority of this from the `Ord` trait.

now getting fancy:

we could do something like,

```rust
/// A trait for types that form a finite total order, with a stable discrete step between adjacent values.
pub trait UniformOrd {
    /// The value such that for an item `n` in the sequence `n, n+1, ..`, `n + STEP == n+1`.
    const STEP: Self;
    const LAST_VALUE: Self;
}
```

and then we could just rely on `Ord` + `Add`, but... this is maybe a bit too fancy?

I would probably instead just do something like,

```rust
A trait for determining adjacency of ordered sets
pub trait OrdAdjacent {
    /// Returns `true` if `rhs` is immediately adjacent to `self` in the ordering.
    fn is_adjacent(self, rhs: Self) -> bool;
}
```

File: read-fonts/src/collections.rs [hunk 2]
```diff
@@ -2,3 +2,6 @@
 
 pub mod int_set;
 pub use int_set::IntSet;
+
+pub mod range_set;
```
Comment (cmyr) [inline read-fonts/src/collections.rs:6]: I'm not sure if we need to make the module public? If there are only a very small number of items we could export them all.

**Review by cmyr — APPROVED** [review]
Some notes inline.

- should we consider whether or not there is an ecosystem crate that suits our needs?

File: read-fonts/src/collections/range_set/mod.rs [hunk 3]
```diff
@@ -0,0 +1,246 @@
...
+use core::{
+    cmp::{max, min},
+    ops::RangeInclusive,
+};
+use std::collections::BTreeMap;
+
+use types::Fixed;
+
+#[derive(Default)]
+pub struct RangeSet<T> {
```
[the remark this answers is not in the document]
Comment (garretrieger) [reply read-fonts/src/collections/range_set/mod.rs:14]: I've got a second PR coming which will add those and some more public api stuff.

File: read-fonts/src/collections/range_set/mod.rs (hunk 1 above)
[the remark this answers is not in the document]
Comment (garretrieger) [reply read-fonts/src/collections/range_set/mod.rs:21]: Yeah exactly, I just need some way to check if two values are adjacent. I think the is_adjacent trait makes sense.

File: read-fonts/src/collections/range_set/mod.rs [hunk 4]
```diff
@@ -0,0 +1,246 @@
...
+
+pub trait Sequence<T> {
+    fn next(&self) -> Option<T>;
+}
+
+impl<T> RangeSet<T>
+where
+    T: Ord + Copy + Sequence<T>,
+{
+    pub fn insert(&mut self, start: T, end: T) {
```
[the remark this answers is not in the document]
Comment (garretrieger) [reply read-fonts/src/collections/range_set/mod.rs:27]: In the follow up PR I've switch to RangeInclusive in the public API. For the use cases I need this for it's sufficient to just work with only inclusive ranges.

[the remark this answers is not in the document]
Comment (garretrieger) [reply read-fonts/src/collections/range_set/mod.rs:27]: Actually I'll just push those changes into this PR

File: read-fonts/src/collections/range_set.rs [hunk 5]
```diff
@@ -0,0 +1,395 @@
...
+/// Overlapping and adjacent ranges are automatically merged together.
+pub struct RangeSet<T> {
+    // an entry in the map ranges[a] = b implies there is an range [a, b] (inclusive) in this set.
+    ranges: BTreeMap<T, T>,
+}
+
+/// Allows a two values to be tested for adjacency.
+pub trait OrdAdjacency<T> {
+    /// Returns true if self is adjacent to rhs
+    fn are_adjacent(self, rhs: T) -> bool;
```
Comment (cmyr) [inline read-fonts/src/collections/range_set.rs:27]: we should clarify if sidedness matters, or if it is strictly about testing increasing order.

File: read-fonts/src/collections/range_set.rs [hunk 6]
```diff
@@ -0,0 +1,395 @@
...
+/// A set of disjoint ranges over numeric types.
+///
+/// Overlapping and adjacent ranges are automatically merged together.
+pub struct RangeSet<T> {
+    // an entry in the map ranges[a] = b implies there is an range [a, b] (inclusive) in this set.
+    ranges: BTreeMap<T, T>,
+}
+
+/// Allows a two values to be tested for adjacency.
+pub trait OrdAdjacency<T> {
```
Comment (cmyr) [inline read-fonts/src/collections/range_set.rs:25]: do we need a generic param here? Is the `T` ever other than `Self`?

File: read-fonts/src/collections/range_set.rs [hunk 7]
```diff
@@ -0,0 +1,395 @@
...
+            write!(f, "[{:?}, {:?}], ", start, end)?;
+        }
+        write!(f, "}}")
+    }
+}
+
+struct IntersectionIter<A, B, T>
+where
+    A: Iterator<Item = RangeInclusive<T>>,
+    B: Iterator<Item = RangeInclusive<T>>,
```
Comment (cmyr) [inline read-fonts/src/collections/range_set.rs:143]: we can delete the bounds here I think, as long as they're on the impl.

**Review by cmyr — APPROVED** [review]
looks good, some notes inline

File: read-fonts/src/collections/range_set.rs (hunk 7 above)
Comment (garretrieger) [reply read-fonts/src/collections/range_set.rs:143]: Peekable requires the types to be iterators.

**Pull request merged**